pkg/runtime/nodejs/nodejs.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	return isNodeJS, nil
+}
+
+func RuntimeInfo(proc procfs.Proc) (*runtime.Runtime, error) {
+	isNodeJS, err := IsRuntime(proc)
+	if err != nil {
+		return nil, fmt.Errorf("is runtime: %w", err)
+	}
+	if !isNodeJS {
+		return nil, nil //nolint:nilnil
+	}
+
+	rt := &runtime.Runtime{
+		Type: "nodejs",
+	}
+
+	exe, err := proc.Executable()
+	if err != nil {
+		return rt, err
+	}
+
+	f, err := os.Open(absolutePath(proc, exe))
+	if err != nil {
+		return rt, fmt.Errorf("open executable: %w", err)
+	}
+	defer f.Close()
+
+	versionString, err := versionFromData(f)
+	if err == nil {
+		version, err := semver.NewVersion(versionString)
+		if err != nil {
+			return rt, fmt.Errorf("new version, %s: %w", versionString, err)
+		}
+		rt.Version = version
+		return rt, nil
+	}
+
+	maps, err := proc.ProcMaps()
+	if err != nil {
+		return nil, fmt.Errorf("error reading process maps: %w", err)
+	}
+
+	var (
+		found bool
+		lib   string
+	)
+	for _, m := range maps {
+		if pathname := m.Pathname; pathname != "" {
+			if m.Perms.Execute {
+				if isNodeJSLib(pathname) {
+					found = true
+					lib = pathname
+					break
+				}
+			}
+		}
+	}
+	if !found {
+		return rt, fmt.Errorf("library %q not found in process maps", exe)
+	}
+
+	lf, err := os.Open(absolutePath(proc, lib))
+	if err != nil {
+		return rt, fmt.Errorf("open library: %w", err)
+	}
+	defer lf.Close()
+
+	versionString, err = versionFromData(lf)
+	if err != nil {
+		return rt, fmt.Errorf("version from data: %w", err)
+	}
+
+	version, err := semver.NewVersion(versionString)
+	if err != nil {
+		return rt, fmt.Errorf("new version, %s: %w", versionString, err)
+	}
+	rt.Version = version
+	return rt, nil
+}
+
+func isNodeJSBin(exe string) bool {
+	return path.Base(exe) == "node" || path.Base(exe) == "nodemon" || path.Base(exe) == "nodejs"
+}
+
+var libRegex = regexp.MustCompile(`/libnode.so\.[0-9]+`)
+
+func isNodeJSLib(lib string) bool {
+	return libRegex.MatchString(lib)
+}
+
+func absolutePath(proc procfs.Proc, p string) string {
+	return path.Join("/proc/", strconv.Itoa(proc.PID), "/root/", p)
+}
+
+func versionFromData(f *os.File) (string, error) {
+	ef, err := elf.NewFile(f)
+	if err != nil {
+		return "", fmt.Errorf("new file: %w", err)
+	}
+	defer ef.Close()
+
+	var lastError error
+	for _, sec := range ef.Sections {
+		if sec.Name == ".data" || sec.Name == ".rodata" {
+			data, err := sec.Data()
+			if err != nil {
+				lastError = fmt.Errorf("read data: %w", err)
+				continue
+			}
+			versionString, err := scanVersionBytes(data)
+			if err != nil {
+				lastError = fmt.Errorf("scan version bytes: %w", err)
+				continue
+			}
+			return versionString, nil
+		}
+	}
+	// If it is found, execution should never reach here.
+	if lastError != nil {
+		return "", lastError
+	}
+	return "", errors.New("version not found")
+}
+
+const semVerRegex string = `v([0-9]+)(\.[0-9]+)(\.[0-9]+)` +
+	`(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?` +
+	`(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?`
```

Comment:
nit: Perhaps removing everything after `-` or `+` and trimming a leading `v` could be easier to understand and evolve? This is a major nit though, I don't mind the approach and you added awesome tests already!

Reply:
We can be more precise about matching, but we can't be loose. Removing `v` gives false matches (I've tested, and we get things like `127.0.0.1). We must be strict since we search the whole space of constant strings. I wish there were a more straightforward way. We could evolve the approach when we know more.